src/server/zset_family.cc

```diff
@@ -870,61 +898,122 @@ OpResult<AddResult> OpAdd(const OpArgs& op_args, const ZParams& zparams, string_
+  ToUpper(&args[arg_pos]);
+  string_view agg = ArgS(args, arg_pos);
+  if (agg == "SUM") {
+    union_args.agg_type = AggType::SUM;
+  } else if (agg == "MIN") {
+    union_args.agg_type = AggType::MIN;
+  } else if (agg == "MAX") {
+    union_args.agg_type = AggType::MAX;
+  } else {
+    return OpStatus::SYNTAX_ERR;
+  }
+  return OpStatus::OK;
+}
+
+OpResult<unsigned> ParseAggregate(UnionArgs& union_args, CmdArgList args, unsigned arg_pos,
+                                  bool store) {
+  if (store && arg_pos + 2 != args.size()) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  if (!store && arg_pos + 3 < args.size()) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  const auto& filled = FillAggType(union_args, args, arg_pos + 1);
+  if (filled == OpStatus::SYNTAX_ERR) {
+    return filled.status();
+  }
+  return arg_pos + 1;
+}
+
+OpResult<unsigned> ParseWeights(UnionArgs& union_args, CmdArgList args, unsigned arg_pos) {
+  if (args.size() <= arg_pos + union_args.num_keys) {
+    return OpStatus::SYNTAX_ERR;
+  }
+
+  for (unsigned j = 0; j < union_args.num_keys; ++j) {
+    string_view weight = ArgS(args, arg_pos + j + 1);
+    if (!absl::SimpleAtod(weight, &(union_args.weights[j]))) {
+      return OpStatus::INVALID_FLOAT;
+    }
+  }
+  return arg_pos + union_args.num_keys;
+}
 
+OpResult<unsigned> ValidateKeyCount(string_view arg_num_keys) {
   // we parsed the structure before, when transaction has been initialized.
-  CHECK(absl::SimpleAtoi(num_str, &store_args.num_keys));
-  DCHECK_GE(args.size(), 3 + store_args.num_keys);
+  unsigned num_keys;
+  if (!absl::SimpleAtoi(arg_num_keys, &num_keys)) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  return num_keys;
+}
+
+OpResult<unsigned> ParseWithScores(UnionArgs& union_args, CmdArgList args, unsigned arg_pos) {
+  if (arg_pos + 1 != args.size()) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  union_args.with_scores = true;
+  return arg_pos;
+}
+
+OpResult<UnionArgs> ParseUnionArgs(CmdArgList args, bool store) {
+  // TODO: support variadic key for ZUNION command (now fixed to 3)
+  string_view num_str = store ? ArgS(args, 2) : "3";
+  UnionArgs union_args;
+
+  const auto& parsed = ValidateKeyCount(num_str);
+  if (parsed == OpStatus::SYNTAX_ERR) {
+    return parsed.status();
+  }
+  union_args.num_keys = *parsed;
+  unsigned key_start = store ? 3 + union_args.num_keys : 1 + union_args.num_keys;
+  // TODO: modify this check when there is variadic key support for ZUNION command
+  DCHECK_GE(args.size(), key_start);
+
+  union_args.weights.resize(union_args.num_keys, 1);
 
-  store_args.weights.resize(store_args.num_keys, 1);
-  for (size_t i = 3 + store_args.num_keys; i < args.size(); ++i) {
+  for (size_t i = key_start; i < args.size(); ++i) {
     ToUpper(&args[i]);
     string_view arg = ArgS(args, i);
     if (arg == "WEIGHTS") {
-      if (args.size() <= i + store_args.num_keys) {
-        return OpStatus::SYNTAX_ERR;
+      const auto& parsed = ParseWeights(union_args, args, i);
```

Comment:
it's wrong to declare `const &` even though it works. also use `args.subspan(i)` - clearer.

Reply:
Modified - part of https://github.com/dragonflydb/dragonfly/pull/420/commits/e44d27b75530be3c97e787eca68957f7b1a3f088